README.md
# ScRNA-seq workflow CONCLUS: from CONsensus CLUSters to a meaningful CONCLUSion


  
  Polina Pavlovich<sup>1,2,3</sup>, Nicolas Descostes<sup>1*</sup>, and Christophe Lancrin<sup>1</sup>
  
  <sup>1</sup>European Molecular Biology Laboratory, EMBL, Rome, Italy <br/>
  <sup>2</sup>Moscow Institute of Physics and Technology, MIPT,  Moscow, Russia <br/>
  <sup>3</sup>Current address: Max Planck Institute of Immunobiology and Epigenetics, MPI-IE, Freiburg, Germany <br/>
  <sup>*</sup>Current maintainer. <br/>
  

A pretty doc version of this vignette is available <a href="http://htmlpreview.github.com/?https://github.com/lancrinlab/CONCLUS/blob/master/vignettes/html_format/conclus_vignette.html" target="_blank"> here </a>. 


# Introduction

CONCLUS is a tool for robust clustering and positive marker features selection of 
single-cell RNA-seq (sc-RNA-seq) datasets. Of note, CONCLUS does not cover the 
preprocessing steps of sequencing files obtained following next-generation 
sequencing. You can find a good resource to start with 
[here](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html). 

CONCLUS is organized into the following steps:

  + Generation of multiple t-SNE plots with a range of parameters including 
    different selection of genes extracted from PCA.
  + Use the Density-based spatial clustering of applications with noise (DBSCAN)
    algorithm for idenfication of clusters in each generated t-SNE plot.
  + All DBSCAN results are combined into a cell similarity matrix.
  + The cell similarity matrix is used to define "CONSENSUS" clusters conserved
    accross the previously defined clustering solutions.
  + Identify marker genes for each concensus cluster.
  

# Getting help

Issues can be submitted directly on github: https://github.com/lancrinlab/CONCLUS/issues. To contact us directly write to christophe.lancrin@embl.it. This package was developed by Polina Pavlovich who is now doing
her Ph.D at the Max Planck Institute of Immunobiology and Epigenetics.


# Standard workflow

## Installation

```{r loading_conclus, results="hide", warning = FALSE, message = FALSE}
# required R >= 3.4.
if(!requireNamespace("devtools"))
	install.packages("devtools")

if(!requireNamespace("BiocManager"))
	install.packages("BiocManager")

library(devtools)
library(BiocManager)

if(!requireNamespace("conclus"))
	install_github("lancrinlab/CONCLUS", dependencies = TRUE)
```

If you get a message such as: *packages ‘BiocParallel’, ‘scran’, ‘scater’, ‘monocle’, ‘SingleCellExperiment’, ‘KEGGREST’, ‘AnnotationDbi’, ‘biomaRt’, ‘org.Mm.eg.db’, ‘S4Vectors’, ‘Biobase’ are not available (for R version 3.6.0)*,
please run the following commands:

```{r, eval = FALSE}
install(c("BiocParallel", "scran", "scater", "monocle", "SingleCellExperiment", 
"KEGGREST", "AnnotationDbi", "biomaRt", "org.Mm.eg.db", "S4Vectors", "Biobase"))
install_github("lancrinlab/CONCLUS", dependencies = TRUE)
```


## Quick start

CONCLUS requires to start with a raw-count matrix with reads or unique molecular 
identifiers (UMIs). The columns of the count matrix must contain cells and the 
rows -- genes. CONCLUS needs a large number of cells to collect statistics, 
we recommend using CONCLUS if you have at least 100 cells.

```{r runConclus_first, results='hide', warning = FALSE, fig.keep = 'none', message= FALSE}
library(conclus)

## Replace by the path to the directory in which results should be written
outputDirectory <- "./YourOutputDirectory"
experimentName <- "Bergiers"

countMatrix <- read.delim(file.path(system.file("extdata", package = "conclus"),		
                                    "Bergiers_counts_matrix_filtered.tsv"), 
                          stringsAsFactors = FALSE)
columnsMetaData <- read.delim(file.path(system.file("extdata", package = "conclus"), 
                                "Bergiers_colData_filtered.tsv"))

sceObjectCONCLUS <- runCONCLUS(outputDirectory, 
                               experimentName,
                               columnsMetaData,
		                       species = "mmu", 
                               plotPDFcellSim = TRUE, # FALSE for > 2500 cells
                               k = 10, cores = 1,
                               statePalette = c("bisque", "cadetblue2", 
                                                "coral1", "cornflowerblue"),
                               deleteOutliers = FALSE)
                               
exportClusteringResults(sceObjectCONCLUS, outputDirectory, experimentName, 
                        "clusters_table.tsv")
```

Five folders will appear in your "outputDirectory": 

  + *output_tables*:
  
    - _cellsSimilarityMatrix.csv: Similarity matrix used to define the consensus clusters using every cells.
    - _clustersSimilarityMatrix.csv: Simplified similarity matrix with median averaged cell values per cluster.
    - _clusters_table.tsv: Cell-cluster corresponding table for defined consensus clusters.
    - _dbscan_results.tsv: Cell-cluster corresponding table for each db-scan parameter.
    - _tSNEResults.rds: Saved RDS object containing a matrix of all t-SNE results (for different PCs and perplexities, see below).  
  
  + *pictures*: This folder contains two similarity matrices clustering heatmaps. One using all cells and the other median averaged by cluster.
                It also contains a subfolder "tSNE_pictures" containing all tSNEs generated with different PCs and perplexity parameters,
                and colored by condition, by cluster found, by dbscan, or uncolored (see below for details).
  + *marker_genes*: This folder contains a list of marker genes, ordered by decreasing score, for each consensus cluster.
  + *tsnes*: Folder containing, for each PCs/complexity values, the tSNE coordinates of each cell.

Further details about how all results are generated can be found below.


## Data

In this vignette, we demonstrate how to use CONCLUS on a sc-RNA-seq dataset from 
[*Bergiers et al. eLife 2018*](https://elifesciences.org/articles/29312). The design for this experiment is described in 
([*Figure 4—figure supplement 2*](https://elifesciences.org/articles/29312/figures#fig4s2)). Bergiers et al. goal was to analyze 
the effect of the simultaneous expression of eight transcription factors (8TFs): *Runx1* - and its partner - *Cbfb*, *Gata2*, 
*Tal1*, *Fli1*, *Lyl1*, *Erg* and *Lmo2* in *in vitro* differentiated embryonic stem cells (ESCs) in comparison to control. 
They knocked-in a polycistronic transgenic construct allowing to over-express eigth transcription factors (i8TFs) simultaneously 
after adding doxycycline (dox). The Empty ESC line did not have any transgene. 
There were **four conditions**: E_minus (Empty no dox), E_plus (Empty with dox), i8TFs_minus (i8TFs no dox) and i8TFs_plus 
(i8TFs with dox). 

This sc-RNA-seq experiment was performed using the SMARTer ICELL8 Single-Cell System ([Click here for more info](https://www.takarabio.com/learning-centers/automation-systems/icell8-introduction/icell8-technology-overview)). The protocol was based on 3' end RNA 
sequencing where each mRNA molecule is labeled with a unique molecular identifier (UMI) during reverse transcription in every 
single cell. The analysis performed by *Bergiers et al.* was based on the dimensionality reduction algorithm called Principal 
Component Analysis (PCA), and they found that there was a major gene expression difference between i8TFs_plus and the other three 
conditions (*Figure 4—figure supplement 2*). **However, it was not clear if other subclusters could be identified consistently in 
this dataset besides the two major clusters. In the current tutorial, we show how CONCLUS can help us to answer this question.**

Labels of the four conditions are in the *state* column of *columnsMetaData*. To avoid a bias in the clustering analysis due to the 
high expression of the eight transcription factors construct, we deleted genes *Cbfb*, *Gata2*, *Tal1*, *Fli1*, *Lyl1*, *Erg*, and *Lmo2* 
from the provided *countMatrix*. Highly abundant embryonic hemoglobins with names starting with "Hba" or "Hbb" were also excluded because 
they seemed to be a primary source of contamination.


## Test clustering

The *TestClustering* function runs one clustering round out of the 84 (default) rounds that CONCLUS normally performs. This step can be useful to determine if the default DBSCAN parameters are suitable for your dataset. 
By default, they are *dbscanEpsilon = c(1.3, 1.4, 1.5)* and *minPts = c(3,4)*. If the dashed horizontal line in the k-NN distance plot lays on the “knee” of the curve (as shown below), it means that optimal epsilon is equal
to the intersection of the line to the y-axis. In our example, optimal epsilon is 1.4 for 5-NN distance where 5 corresponds to MinPts. 

In the "test_clustering" folder under outputDirectory, the three plots below will be saved where one corresponds to the “distance_graph.pdf” (Figure1), another one to “test_tSNE.pdf” (*p[[1]]*, Figure2), and the last one will be saved as 
“test_clustering.pdf” (*p[[3]]*, Figure3).


```{r normalization_1, results="hide", warning = FALSE}
sceObject <- conclus::normaliseCountMatrix(countMatrix, species = "mmu", 
                                  colData = columnsMetaData)
```


```{r testClustering, fig.height=5, fig.width=6, fig.keep = 'none'}
p <- conclus::testClustering(sceObject, outputDirectory, experimentName)
```

<img src="vignettes/figures/fig1_KNN.png"
     alt="k-NN distance plot"
     style="float: left; margin-right: 10px;" />


```{r testClustering_result1, eval = FALSE}
# saved as "outputDirectory/test_clustering/test_tSNE.pdf"
p[[1]]
```

<img src="vignettes/figures/fig2_tSNEnoColor.png"
     alt="k-NN distance plot"
     style="float: left; margin-right: 10px;" />



```{r testClustering_result2, eval = FALSE}
# saved as "outputDirectory/test_clustering/test_clustering.pdf"
p[[3]]
```

<img src="vignettes/figures/fig3_tSNEdbScan.png"
     alt="k-NN distance plot"
     style="float: left; margin-right: 10px;" />


# CONCLUS step by step

The wrapper function runCONCLUS is organized into 7 steps:

  + Normalization of the counts matrix
  + Generation of t-SNE coordinates
  + Clustering with DB-SCAN
  + Cell and cluster similarity matrix calculation
  + Plotting
  + Results export
  + Marker genes identification


## Normalization of the counts matrix

sc-RNA-seq datasets are quite challenging notably because of sparsity (many genes are not detected consistently yielding expression matrices with many zeroes) and also because of technical noise. 
To facilitate analysis, one needs to perform a step of normalization which allows the correction of unwanted technical and biological noises (click [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5549838/) for a complete review on
normalization techniques).

CONCLUS uses Scran and Scater packages for normalization. Beforehand, the function will annotate genes creating *rowData* and add statistics about cells into *columnsMetaData*. If you already have *columnsMetaData* and *rowData*, 
you can give it to the function (see manual). It will keep your columns and add new ones at the end. If you do not want to lose any cell after quality metrics check, select *alreadyCellFiltered = TRUE*, by default it is *FALSE*. 
Before *scran* and *scater* normalization, the function will call *scran::quickCluster* (see manual for details). If you want to skip this step, type *runQuickCluster = FALSE*, by default it is *TRUE*. 
We recommend to use *runQuickCluster = TRUE* for medium-size datasets with 500-10000 cells. However, it can take a significant amount of time for a larger amount of cells and will not be useful for small sets of 200-300 samples. 

This function is currently limited to human and mouse.
\newline
```{r normalization_2, results="hide", warning = FALSE, message = FALSE}
sceObject <- conclus::normaliseCountMatrix(countMatrix, species = "mmu", 
                                  colData = columnsMetaData)
```

The function *normaliseCountMatrix* returns a *sceObject* with slots *counts*, *exprs*, *colData*, and *rowData*. *rowData* keeps information about GO terms containing genes on the cell surface or involved in secretion. 
This information can help to study cross-talk between cell types or find surface protein-coding marker genes suitable for flow cytometry. The columns with the GO terms are *go_id* and *name_1006* (see manual).

The slots can be accessed as indicated below:
\newline
```{r normalization_results}
# checking what changed after the normalisation
dim(sceObject)

# show first columns and rows of the count matrix
SingleCellExperiment::counts(sceObject)[1:5,1:5]

# show first columns and rows of the normalized count matrix
Biobase::exprs(sceObject)[1:5,1:5]

# visualize first rows of metadata (coldata)
coldataSCE <- as.data.frame(SummarizedExperiment::colData(sceObject))
head(coldataSCE)

# visualize beginning of the rowdata containing gene information
rowdataSCE <- as.data.frame(SummarizedExperiment:::rowData(sceObject))
head(rowdataSCE)
```


## Generation of t-SNE coordinates

*runCONCLUS* creates needed output folders (if you did not run *testClustering* beforehand). Then it generates an object of fourteen (by default) tables with tSNE coordinates. Fourteen because it will vary seven values of 
principal components *PCs=c(4, 6, 8, 10, 20, 40, 50)* and two values of perplexity *perplexities=c(30, 40)* in all possible combinations. 

The chosen values of PCs and perplexities can be changed if necessary. We found that this combination works well for sc-RNA-seq datasets with 400-2000 cells. If you have 4000-9000 cells and expect more than 15 clusters, 
we recommend to use more first PCs and higher perplexity, for example, *PCs=c(8, 10, 20, 40, 50, 80, 100)* and *perplexities=c(200, 240)*. For details about perplexities parameter see ‘?Rtsne’.
\newline
```{r tsne_generation, results="hide", warning = FALSE}
initialisePath(outputDirectory)
# default parameters, can be selected by a user
PCs=c(4, 6, 8, 10, 20, 40, 50)
perplexities=c(30, 40)
randomSeed = 42
tSNEResults <- generateTSNECoordinates(sceObject, outputDirectory, 
                                          experimentName, PCs=PCs, 
                                          perplexities=perplexities,
                                          randomSeed = randomSeed)
```

Results can be explored as follows:
\newline
```{r tsne_showResults}
ncol(tSNEResults)
# the third matrix of t-SNE coordinates with PC = 8 and perplixities = 30
# it is saved as "tsnes/Bergiers_tsne_coordinates_3_8PCs_30perp.tsv"
head(tSNEResults[1,3][[1]])
```


## Clustering with DB-SCAN

Following the calculation of t-SNE coordinates, DBSCAN is run with a range of epsilon and MinPoints values which will yield a total of 84 clustering solutions (PCs x perplexities x MinPoints x epsilon). *minPoints* is the minimum 
cluster size which you assume to be meaningful for your experiment and *epsilon* is the radius around the cell where the algorithm will try to find *minPoints* dots. Optimal *epsilon* must lay on the knee of the k-NN function as 
shown in the "test_clustering/distance_graph.pdf" (See Test clustering section above). 
\newline
```{r runDBSScan, results="hide", warning = FALSE}
epsilon=c(1.3, 1.4, 1.5)
minPoints=c(3, 4)
cores=1
dbscanResults <- conclus::runDBSCAN(tSNEResults, sceObject, outputDirectory, 
                           experimentName, epsilon=epsilon, 
                           minPoints=minPoints,
                           cores=cores)
```


## Cell and cluster similarity matrix calculation

The above calculated results are combined together in a matrix called “cell similarity matrix”. *RunDBSCAN* function returns a matrix where columns are cells and rows are 84 clustering solutions (which is equal to number of 
PCs x perplexities x MinPoints x epsilon). Since the range of cluster varies from result to result, there is no exact match between numbers in different rows of the returned matrix. Cells having the same number within a row are 
guaranteed to be in one cluster. We can calculate how many times out of 84 clustering solutions, every two cells were in one cluster and that is how we come to the similarity matrix of cells. We want to underline that zero in the 
*dbscanResults* means that a cell was not assigned to any cluster, so two cells with zeros were not necessarily similar, that is why we count clusters starting from one. *clusterCellsInternal* is a general function which returns a 

NEWS
## 0.99.3 (Nicolas Descostes)

- Correct bug in runCONCLUS to handle manual clustering
- Update vignette
- Provide vignette on welcome page of github repository and link it to the prettydoc version

## 0.99.2 (Nicolas Descostes)

- Project moved to lancrinlab account

## 0.99.1 (Nicolas Descostes)

- Modification of the getTSNEresults to include updates on scater package

- Include missing packages into NAMESPACE file

- Re-structuration and modifications of the vignette

- Add normalization step to runCONCLUS function

- Add cores parameter to calls of generateTSNECoordinates

- Hide getTSNEresults, plotDistanceGraphWithEpsilon, plotTestClustering, mkDbscan, 
mkOutlierScoreDf, excludeOutliers, mkSimMat, mkSimMed, orderCellsInCluster, 
orderGenesInCluster, generateAnnotationColors, annotateGenes, filterCells, 
addCellsInfo, runClustering and filterGenes functions

- Remove deleteFile, checkTSNEPicture, getKEGGGenes, chooseStatePalette, 
plotDistanceGraph functions

- Reorganize import and exports in NameSpace

- Suppress packages loading from R code and put them in NAMESPACE

- Delete runClustering.Rd

- Modification of clusterCellsInternal.Rd, runCONCLUS.Rd, normaliseCountMatrix.Rd, 
generateTSNECoordinates.Rd, runDBSCAN.Rd, calculateClustersSimilarity.Rd, 
plotCellSimilarity.Rd, plotClusteredTSNE.Rd, exportMatrix.Rd, exportData.Rd,
exportClusteringResults.Rd, rankGenes.Rd, getMarkerGenes.Rd, plotCellHeatmap.Rd,
plotGeneExpression.Rd, markers.Rd, getGenesInfo.Rd, saveMarkersLists.Rd,   
saveGenesInfo.Rd, addClusteringManually.Rd, choosePalette.Rd, initialisePath.Rd,
plotClustersSimilarity.Rd

- Remove roxygen documentation from the code

- Move .getTSNEresults, choosePalette, initialisePath to sharedInternals.R

- Move testClustering, plotDistanceGraphWithEpsilon, plotTestClustering,
excludeOutliers, mkOutlierScoreDf, runClustering, mkSimMat, clusterCellsInternal,
mkSimMed, calculateClustersSimilarity, addClusteringManually to clustering.R

- Move runCONCLUS function to runCONCLUS.R

- Move exportMatrix, exportData, exportClusteringResults to export.R

- Move plotCellSimilarity, plotClusteredTSNE, plotCellHeatmap, 
plotGeneExpression, plotClustersSimilarity, orderCellsInCluster, orderGenesInCluster,
generateAnnotationColors  to plot.R

- Move generateTSNECoordinates to tSNE.R

- Move mkDbscan and runDBSCAN to dbScan.R

- Move rankGenes, rankGenesInternal, getMarkerGenes, getGenesInfo, saveGenesInfo, 
saveMarkersLists to markers.R

- Move normaliseCountMatrix, normaliseCountMatrix, duplicate, annotateGenes, 
filterCells, addCellsInfo, filterGenes to normalization.R

- Merge the three normalization function in normalization.R by setting 
databaseDir = TRUE and modifying annotateGene accordingly

- Merge .plotClusteredTSNE and plotClusteredTSNE

- Merge .plotGeneExpression and plotGeneExpression

- Merge rankGenes and rankGenesInternal

- Simplify testClustering for pdf parameters

- Remove first call of testClustering and keep internal function code only

- Set one core by default in runCONCLUS.R, tSNE.R, dbScan.R

- Modify .choosePalette in sharedInternal.R to avoid repetition of colors

- Remove the method parameter from normaliseCountMatrix in normalization.R

- Change ordercluster in plotCellSimilarity (in plot.R) to take the values TRUE/FALSE instead of name/FALSE.

- Change ordercluster in plotCellHeatmap (in plot.R) to take the values TRUE/FALSE instead of name/FALSE. Add an additional "similarity" parameter.

- Add default value for clusteringMethod param of calculateClustersSimilarity in clustering.R

- Change default values of clusterCellsInternal in clustering.R

- Add plotClustersSimilarity default values in plot.R

- Add param value checking for columnName param in plotClusteredTSNE in plot.R

- Move detailed explanation of rankGenes in the vignette to the "details" section of rankGenes.Rd

- Reorganize parameters for getMarkerGenes and set default gene number to 10 in markers.R

- Reorganize parameters of plotGeneExpression in plot.R

- Add DataCombine to namespace

- Replace coresGene parameter by cores and set default to 1 in markers.R

- change how results of clusterCellsInternal are handled if preClustered is TRUE in runCONCLUS.R

- Add a manualClusteringObject parameter to handle the case in which the clusters were modified with addClusteringManually in runCONCLUS.R

- change columnsMetaData and species to optional parameters in runCONCLUS.R

- set test on manualClusteringObject and preClustered parameters in runCONCLUS.R

- Add Descostes name to NAMESPACE



toDelete_dev/images_export.R

# Quick start

library(conclus)

outputDirectory <- "/home/descostes/Documents/analysis/lancrin/conclus/vignette_images/"
experimentName <- "Bergiers"

countMatrix <- read.delim(file.path(system.file("extdata", package = "conclus"),
				"Bergiers_counts_matrix_filtered.tsv"), 
		stringsAsFactors = FALSE)
columnsMetaData <- read.delim(file.path(system.file("extdata", package = "conclus"), 
				"Bergiers_colData_filtered.tsv"))

sceObjectCONCLUS <- runCONCLUS(outputDirectory, 
		experimentName,
		columnsMetaData,
		species = "mmu", 
		plotPDFcellSim = TRUE, # FALSE for > 2500 cells
		k = 10, cores = 1,
		statePalette = c("bisque", "cadetblue2", 
				"coral1", "cornflowerblue"),
		deleteOutliers = FALSE)

exportClusteringResults(sceObjectCONCLUS, outputDirectory, experimentName, 
		"clusters_table.tsv")

## Test clustering


sceObject <- conclus::normaliseCountMatrix(countMatrix, species = "mmu", 
		colData = columnsMetaData)


png(paste0(outputDirectory, "fig1_KNN.png"), width = 700, height=700, res=100)
p <- conclus::testClustering(sceObject, outputDirectory, experimentName)
dev.off()

png(paste0(outputDirectory, "fig2_tSNEnoColor.png"), width = 700, height=700, res=100, type = "cairo-png")
p[[1]]
dev.off()

png(paste0(outputDirectory, "fig3_tSNEdbScan.png"), width = 700, height=700, res=100, type = "cairo-png")
p[[3]]
dev.off()


# CONCLUS step by step

## Normalization of the counts matrix

sceObject <- conclus::normaliseCountMatrix(countMatrix, species = "mmu", 
		colData = columnsMetaData)

dim(sceObject)
SingleCellExperiment::counts(sceObject)[1:5,1:5]
Biobase::exprs(sceObject)[1:5,1:5]
coldataSCE <- as.data.frame(SummarizedExperiment::colData(sceObject))
head(coldataSCE)
rowdataSCE <- as.data.frame(SummarizedExperiment:::rowData(sceObject))
head(rowdataSCE)


## Generation of t-SNE coordinates

initialisePath(outputDirectory)
# default parameters, can be selected by a user
PCs=c(4, 6, 8, 10, 20, 40, 50)
perplexities=c(30, 40)
randomSeed = 42
tSNEResults <- generateTSNECoordinates(sceObject, outputDirectory, 
		experimentName, PCs=PCs, 
		perplexities=perplexities,
		randomSeed = randomSeed)

ncol(tSNEResults)
head(tSNEResults[1,3][[1]])


## Clustering with DB-SCAN

epsilon=c(1.3, 1.4, 1.5)
minPoints=c(3, 4)
cores=1
dbscanResults <- conclus::runDBSCAN(tSNEResults, sceObject, outputDirectory, 
		experimentName, epsilon=epsilon, 
		minPoints=minPoints,
		cores=cores)

## Cell and cluster similarity matrix calculation

clusteringMethod="ward.D2"
k=10 # parameter for cutree
deepSplit = 0 # 0 to avoid cutreeDynamic, 1 to 4 to use it
message("Calculating cells similarity matrix.")
clusteringResults <- conclus::clusterCellsInternal(dbscanResults, sceObject, clusterNumber=k, 
		deepSplit=deepSplit, cores=cores,
		clusteringMethod=clusteringMethod)
sceObjectFiltered <- clusteringResults[[1]]
cellsSimilarityMatrix <- clusteringResults[[2]]

print(table(SummarizedExperiment::colData(sceObjectFiltered)$clusters, 
				dnn=list("Cells distribuion by clusters")))

clustersSimilarityMatrix <- 
		conclus::calculateClustersSimilarity(cellsSimilarityMatrix, 
				sceObject = sceObjectFiltered,
				clusteringMethod = "ward.D2")[[1]]


## Plotting


### t-SNE colored by clusters or conditions

tSNEclusters <- conclus::plotClusteredTSNE(sceObjectFiltered, outputDirectory, 
		experimentName, PCs=PCs, perplexities=perplexities, 
		colorPalette = "default", columnName = "clusters", 
		returnPlot = TRUE)

tSNEnoColor <- conclus::plotClusteredTSNE(sceObjectFiltered, outputDirectory, 
		experimentName, PCs=PCs, perplexities=perplexities, 
		colorPalette = "default", columnName = "noColor", 
		returnPlot = TRUE)

if(any(colnames(SummarizedExperiment::colData(sceObjectFiltered)) 
				%in% "state")){
	tSNEstate <- conclus::plotClusteredTSNE(sceObjectFiltered, outputDirectory, 
			experimentName, PCs=PCs, perplexities=perplexities, 
			colorPalette = "default", columnName = "state", 
			returnPlot = TRUE)
}

png(paste0(outputDirectory, "fig4_tSNEColor_clusters.png"), width = 700, height=700, res=100, type = "cairo-png")
tSNEclusters[[5]]
dev.off()

png(paste0(outputDirectory, "fig5_tSNEnoColor.png"), width = 700, height=700, res=100, type = "cairo-png")
tSNEnoColor[[5]]
dev.off()

png(paste0(outputDirectory, "fig6_tSNEState.png"), width = 700, height=700, res=100, type = "cairo-png")
tSNEstate[[5]]
dev.off()



### Cell similarity heatmap


colorPalette="default"
statePalette="default"
plotPDFcellSim = TRUE
orderClusters = FALSE
clustersNumber <- length(unique(
				SummarizedExperiment::colData(sceObjectFiltered)$clusters))
colorPalette <- conclus::choosePalette(colorPalette, clustersNumber)

png(paste0(outputDirectory, "fig7_plotCellSimilarity.png"), width = 700, height=700, res = 100)
conclus::plotCellSimilarity(sceObjectFiltered, cellsSimilarityMatrix, outputDirectory,
		experimentName, colorPalette, 
		orderClusters = orderClusters, 
		statePalette = statePalette, 
		clusteringMethod = clusteringMethod,
		plotPDF = plotPDFcellSim,
		returnPlot = TRUE)
dev.off()

### Cluster similarity heatmap

png(paste0(outputDirectory, "fig8_plotClustersSimilarity.png"), width = 700, height=700, res=100)
conclus::plotClustersSimilarity(clustersSimilarityMatrix, 
		sceObjectFiltered,
		dataDirectory = outputDirectory, 
		experimentName = experimentName, 
		colorPalette = colorPalette,
		statePalette = statePalette,
		clusteringMethod = clusteringMethod,
		returnPlot = TRUE)
dev.off()

## Results export

conclus::exportMatrix(cellsSimilarityMatrix, outputDirectory, experimentName, 
		"cellsSimilarityMatrix")
conclus::exportMatrix(clustersSimilarityMatrix, outputDirectory, experimentName, 
		"clustersSimilarityMatrix")
conclus::exportData(sceObjectFiltered, outputDirectory, experimentName)


## Marker genes identification

conclus::rankGenes(sceObjectFiltered, clustersSimilarityMatrix, outputDirectory, 
		experimentName)
rankedGenesClus5 <- read.delim(file.path(outputDirectory, "marker_genes",
				"Bergiers_cluster_5_genes.tsv"),
		stringsAsFactors = FALSE)
head(rankedGenesClus5, n = 10)


# Plot a heatmap with positive marker genes

genesNumber <- 10
markersClusters <- conclus::getMarkerGenes(outputDirectory, sceObjectFiltered, 
		experimentName = experimentName,
		genesNumber = genesNumber)

orderClusters <- T # F will apply hierarchical clustering to all cells
orderGenes <- T    # F will apply hierarchical clustering to all genes
meanCentered <- T  # F to show normalized counts

png(paste0(outputDirectory, "fig9_positivemarkers.png"), width = 700, height=700, res = 160)
conclus::plotCellHeatmap(markersClusters, sceObjectFiltered, outputDirectory, 
		experimentName, 
		paste0("clusters",
				length(levels(
								SummarizedExperiment::colData(sceObjectFiltered)$clusters)),
				"_meanCentered",meanCentered,
				"_orderClusters",orderClusters,
				"_orderGenes",orderGenes,"_top",
				genesNumber, "markersPerCluster"), 
		meanCentered = meanCentered, 
		colorPalette = RColorBrewer::brewer.pal(10, "Paired"),
		orderClusters = orderClusters,
		orderGenes = orderGenes,
		fontsize_row = 4,
		statePalette = c("bisque", "cadetblue2", 
				"coral1", "cornflowerblue"),
		color = colorRampPalette(c("#023b84","#4b97fc", 
						"#FEE395", 
						"#F4794E", "#D73027",
						"#a31008","#7a0f09"))(100),
		returnPlot = TRUE,
		width = 7.5, height = 6.5)
dev.off()



orderClusters <- T # F will apply hierarchical clustering to all cells
orderGenes <- T    # F will apply hierarchical clustering to all genes
meanCentered <- F  # F to show normalized counts

png(paste0(outputDirectory, "fig10_positivemarkers-unnorm.png"), width = 700, height=700, res = 160)
conclus::plotCellHeatmap(markersClusters, sceObjectFiltered, outputDirectory, 
		experimentName, 
		paste0("clusters",
				length(levels(SummarizedExperiment::colData(sceObjectFiltered)$clusters)),
				"_meanCentered",meanCentered,
				"_orderClusters",orderClusters,
				"_orderGenes",orderGenes,"_top",
				genesNumber, "markersPerCluster"), 
		meanCentered = meanCentered, 
		colorPalette = RColorBrewer::brewer.pal(10, "Paired"),
		orderClusters = orderClusters,
		orderGenes = orderGenes,
		fontsize_row = 4,
		statePalette = c("bisque", "cadetblue2", 
				"coral1", "cornflowerblue"),
		color = colorRampPalette(c("#023b84","#4b97fc", 
						"#FEE395", 
						"#F4794E", "#D73027",
						"#a31008","#7a0f09"))(100),
		returnPlot = TRUE)
dev.off()



# Plot t-SNE colored by expression of a selected gene

png(paste0(outputDirectory, "fig11_positivemarkers-ccl3.png"), width = 700, height=700, res = 100)
plotGeneExpression("Ccl3", experimentName, outputDirectory, sceObjectFiltered, 
		tSNEpicture = 10, returnPlot=T)
dev.off()


png(paste0(outputDirectory, "fig12_positivemarkers-Gp9.png"), width = 700, height=700, res = 100)
plotGeneExpression("Gp9", experimentName, outputDirectory, sceObjectFiltered, 
		tSNEpicture = 10, returnPlot=T)
dev.off()

png(paste0(outputDirectory, "fig13_positivemarkers-Fn1.png"), width = 700, height=700, res = 100)
plotGeneExpression("Fn1", experimentName, outputDirectory, sceObjectFiltered, 
		tSNEpicture = 10, returnPlot=T)
dev.off()

png(paste0(outputDirectory, "fig14_positivemarkers-Alox5ap.png"), width = 700, height=700, res = 100)
plotGeneExpression("Alox5ap", experimentName, outputDirectory, sceObjectFiltered, 
		tSNEpicture = 10, returnPlot=T)
dev.off()


# Collect publicly available info about marker genes

## Collect information for the top 10 markers for each cluster

result <- getGenesInfo(markersClusters, groupBy = "clusters",
		getUniprot = TRUE)

outputDir <- file.path(outputDirectory, "/marker_genes/getGenesInfo")
dir.create(outputDir, showWarnings=F)
write.table(result, file = file.path(outputDir, 
				"Bergiers_markersClusters_top10_clusters9_genesInfo.csv"),
		quote = FALSE, sep = ";", row.names = FALSE)


## Collect information for the top X markers for each cluster

saveMarkersLists(experimentName, outputDirectory)
saveGenesInfo(outputDirectory, sep = ";", header = TRUE, 
		startFromFile = 10, getUniprot = TRUE)


# Supervised clustering 

exportClusteringResults(sceObjectFiltered, outputDirectory, experimentName, 
		"clusters_table.tsv")
clustersTable <- read.delim(file.path(outputDirectory, "output_tables", 
				paste0(experimentName, "_clusters_table.tsv")), stringsAsFactors = FALSE) 

#replace “10” by “9” in the clusters_table to merge 9/10
clustersTable$clusters[clustersTable$clusters == "10"] = "9"

write.table(clustersTable, file.path(outputDirectory, "output_tables", 
				paste0(experimentName, "_clusters_table_manual.tsv")), 
		quote = FALSE, sep = "\t")



sceObjectFiltered <- addClusteringManually(fileName = "clusters_table_manual.tsv", 
		dataDirectory = outputDirectory, 
		experimentName = experimentName,
		sceObject = sceObjectFiltered, 
		columnName = "clusters")

# Redo the analysis with manual clustering
sceObjectFiltered <- runCONCLUS(outputDirectory, experimentName, 
		statePalette= c("bisque", "cadetblue2", "coral1", "cornflowerblue"),
		preClustered = TRUE, manualClusteringObject = sceObjectFiltered, cores=10)



meanCentered <- T  # F to show normalized counts, 
png(paste0(outputDirectory, "fig15_plotCellHeatmapMeanCentered_manual.png"), width = 700, height=700, res = 150)
plotCellHeatmap(markersClusters, sceObjectFiltered, outputDirectory, 
		experimentName, 
		paste0("clusters",
				length(levels(
								SummarizedExperiment::colData(sceObjectFiltered)$clusters)),
				"_meanCentered",meanCentered,"_orderClusters",orderClusters,
				"_orderGenes",orderGenes,"_top",
				genesNumber, "markersPerCluster"), 
		meanCentered = meanCentered, 
		colorPalette = c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C",
				"#FB9A99", "#E31A1C", "#FDBF6F", "#FF7F00", "#CAB2D6"),
		orderClusters = orderClusters,
		orderGenes = orderGenes,
		fontsize_row = 5,
		statePalette = c("bisque", "cadetblue2", 
				"coral1", "cornflowerblue"),
		color = colorRampPalette(c("#023b84","#4b97fc", 
						"#FEE395", 
						"#F4794E", "#D73027",
						"#a31008","#7a0f09"))(100),
		returnPlot = TRUE)
dev.off()


		

png(paste0(outputDirectory, "fig16_tSNE_manual_Cd52.png"), width = 700, height=700, res = 100)
plotGeneExpression("Cd52", experimentName, outputDirectory, sceObjectFiltered,
		tSNEpicture = 10, returnPlot = TRUE)
dev.off()

png(paste0(outputDirectory, "fig17_tSNE_manual_Alox5ap.png"), width = 700, height=700, res = 100)
plotGeneExpression("Alox5ap", experimentName, outputDirectory, sceObjectFiltered,
		tSNEpicture = 10, returnPlot = TRUE)
dev.off()
